axum-core/src/response/into_response.rs

```diff
@@ -0,0 +1,507 @@
+/// };
+///
+/// enum MyError {
+///     SomethingWentWrong,
+///     SomethingElseWentWrong,
+/// }
+///
+/// impl IntoResponse for MyError {
+///     fn into_response(self) -> Response {
+///         let body = match self {
+///             MyError::SomethingWentWrong => "something went wrong",
+///             MyError::SomethingElseWentWrong => "something else went wrong",
+///         };
+///
+///         // its often easier to implement `IntoResponse` by calling other implementations
+///         (StatusCode::INTERNAL_SERVER_ERROR, body).into_response()
+///     }
+/// }
+///
+/// // `Result<impl IntoResponse, MyError>` can now be returned from handlers
+/// let app = Router::new().route("/", get(handler));
+///
+/// async fn handler() -> Result<(), MyError> {
+///     Err(MyError::SomethingWentWrong)
+/// }
+/// # async {
+/// # hyper::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+/// # };
+/// ```
+///
+/// Or if you have a custom body type you'll also need to implement
+/// `IntoResponseParts` for it:
+///
+/// ```rust
+/// use axum::{
+///     body,
+///     routing::get,
+///     response::{IntoResponse, Response},
+///     Router,
+/// };
+/// use http_body::Body;
+/// use http::HeaderMap;
+/// use bytes::Bytes;
+/// use std::{
+///     convert::Infallible,
+///     task::{Poll, Context},
+///     pin::Pin,
+/// };
+///
+/// struct MyBody;
+///
+/// // First implement `Body` for `MyBody`. This could for example use
+/// // some custom streaming protocol.
+/// impl Body for MyBody {
+///     type Data = Bytes;
+///     type Error = Infallible;
+///
+///     fn poll_data(
+///         self: Pin<&mut Self>,
+///         cx: &mut Context<'_>
+///     ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+///         # unimplemented!()
+///         // ...
+///     }
+///
+///     fn poll_trailers(
+///         self: Pin<&mut Self>,
+///         cx: &mut Context<'_>
+///     ) -> Poll<Result<Option<HeaderMap>, Self::Error>> {
+///         # unimplemented!()
+///         // ...
+///     }
+/// }
+///
+/// // Now we can implement `IntoResponse` directly for `MyBody`
+/// impl IntoResponse for MyBody {
+///     fn into_response(self) -> Response {
+///         Response::new(body::boxed(self))
+///     }
+/// }
+///
+/// // `MyBody` can now be returned from handlers.
+/// let app = Router::new().route("/", get(|| async { MyBody }));
+/// # async {
+/// # hyper::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+/// # };
+/// ```
+pub trait IntoResponse {
+    /// Create a response.
+    fn into_response(self) -> Response;
+}
+
+impl IntoResponse for StatusCode {
+    fn into_response(self) -> Response {
+        let mut res = ().into_response();
+        *res.status_mut() = self;
+        res
+    }
+}
+
+impl IntoResponse for Version {
+    fn into_response(self) -> Response {
+        let mut res = ().into_response();
+        *res.version_mut() = self;
+        res
+    }
+}
```

Comment:
Kinda surprised no one has asked by `Version` didn't implement `IntoResponse` earlier 🤔

Reply:
Why would you ever want to override that? I feel like the http version should always be set by hyper.

Reply:
Not sure that you need to actually. But tonic does set it, although I think hyper would handle that as well.